File: cpp/src/terminal/TerminalBuffer.cpp
#include "ui/terminal/TerminalBuffer.hpp"
#include "ui/core/TextVisual.hpp"

#include <algorithm>
#include <cstdlib>
#include <sstream>
#include <stdexcept>

namespace ui {

namespace {

constexpr int kAnsiBaseRow = 1;
constexpr int kAnsiBaseCol = 1;
constexpr int kAnsiDefaultForeground = 39;
constexpr int kAnsiDefaultBackground = 49;
constexpr char kAnsiSyncUpdateBegin[] = "\x1b[?2026h";
constexpr char kAnsiSyncUpdateEnd[] = "\x1b[?2026l";

[[nodiscard]] std::size_t cellCountFor(int width, int height) {
    if (width < 0 || height < 0) {
        throw std::invalid_argument("terminal buffer dimensions must not be negative");
    }
    return static_cast<std::size_t>(width * height);
}

[[nodiscard]] std::string sgrForColor(const std::optional<Color>& color, bool foreground) {
    if (!color.has_value() || color->kind() == Color::Kind::Unset || color->isTransparent()) {
        return std::to_string(foreground ? kAnsiDefaultForeground : kAnsiDefaultBackground);
    }
    if (color->rgba().has_value()) {
        const Rgba& rgba = *color->rgba();
        std::ostringstream out;
        out << (foreground ? "38" : "48") << ";2;"
            << static_cast<int>(rgba.red) << ';'
            << static_cast<int>(rgba.green) << ';'
            << static_cast<int>(rgba.blue);
        return out.str();
    }
    return std::to_string(foreground ? kAnsiDefaultForeground : kAnsiDefaultBackground);
}

[[nodiscard]] std::string sgrForCell(const TerminalCell& cell) {
    return "\x1b[" + sgrForColor(cell.foreground, true) + ";" +
           sgrForColor(cell.background, false) + "m";
}

[[nodiscard]] std::string terminalCellText(const std::string& text) {
    return safeTerminalCellText(text);
}

}  // namespace

TerminalBuffer::TerminalBuffer(int width, int height)
    : width_(width),
      height_(height),
      cells_(cellCountFor(width, height)),
      previous_(cells_) {
}

void TerminalBuffer::resize(int width, int height) {
    if (width < 0 || height < 0) {
        throw std::invalid_argument("terminal buffer dimensions must not be negative");
    }
    if (width == width_ && height == height_) {
        return;
    }
    width_ = width;
    height_ = height;
    cells_.assign(cellCountFor(width_, height_), TerminalCell{});
    previous_.assign(cells_.size(), TerminalCell{});
    forceFullRedraw_ = true;
}

void TerminalBuffer::clear(TerminalCell cell) {
    std::ranges::fill(cells_, std::move(cell));
}

void TerminalBuffer::setCell(int row, int col, TerminalCell cell) {
    if (!inBounds(row, col)) {
        return;
    }
    cells_[index(row, col)] = std::move(cell);
}

void TerminalBuffer::requestFullRedraw() {
    forceFullRedraw_ = true;
}

void TerminalBuffer::beginRenderStats() {
    renderStats_ = TerminalBufferRenderStats{};
}

const TerminalCell& TerminalBuffer::cell(int row, int col) const {
    if (!inBounds(row, col)) {
        throw std::out_of_range("terminal cell is out of bounds");
    }
    return cells_[index(row, col)];
}

TerminalBufferRenderStats TerminalBuffer::renderStats() const {
    return renderStats_;
}

std::string TerminalBuffer::renderDiff(int rowOffset, int colOffset) {
    return renderDiffRegion(rowOffset, colOffset, 0, 0, height_, width_);
}

std::string TerminalBuffer::renderDiffRegion(
    int rowOffset,
    int colOffset,
    int startRow,
    int startCol,
    int height,
    int width
) {
    std::string output;
    const bool fullRedraw = forceFullRedraw_;
    bool synchronizeUpdate = false;
    bool rawEmitted = false;
    renderStats_.fullRedraw = renderStats_.fullRedraw || fullRedraw;
    const int firstRow = std::max(0, startRow);
    const int firstCol = std::max(0, startCol);
    const int lastRow = std::min(height_, startRow + std::max(0, height));
    const int lastCol = std::min(width_, startCol + std::max(0, width));
    for (int row = firstRow; row < lastRow; ++row) {
        int col = firstCol;
        while (col < lastCol) {
            const std::size_t startIndex = index(row, col);
            if (cells_[startIndex].rawSkip) {
                if (fullRedraw || cells_[startIndex] != previous_[startIndex]) {
                    ++renderStats_.changedCells;
                }
                previous_[startIndex] = cells_[startIndex];
                ++col;
                continue;
            }
            if (!fullRedraw && cells_[startIndex] == previous_[startIndex]) {
                ++col;
                continue;
            }

            const TerminalCell styleCell = cells_[startIndex];
            if (!styleCell.raw.empty()) {
                synchronizeUpdate = true;
                const int rawWidth = std::max(1, styleCell.rawWidth);
                const int rawHeight = std::max(1, styleCell.rawHeight);
                const int clearWidth = std::min(rawWidth, width_ - col);
                const int clearHeight = std::min(rawHeight, height_ - row);
                for (int clearRow = row; clearRow < row + clearHeight; ++clearRow) {
                    output += "\x1b[" + std::to_string(clearRow + rowOffset + kAnsiBaseRow) + ";" +
                              std::to_string(col + colOffset + kAnsiBaseCol) + "H";
                    output += sgrForCell(styleCell);
                    output.append(static_cast<std::size_t>(clearWidth), ' ');
                }
                if (clearHeight >= rawHeight) {
                    output += "\x1b[" + std::to_string(row + rowOffset + kAnsiBaseRow) + ";" +
                              std::to_string(col + colOffset + kAnsiBaseCol) + "H";
                    output += styleCell.raw;
                    rawEmitted = true;
                }
                for (int coveredRow = row; coveredRow < row + clearHeight; ++coveredRow) {
                    for (int coveredCol = col; coveredCol < col + clearWidth; ++coveredCol) {
                        previous_[index(coveredRow, coveredCol)] = cells_[index(coveredRow, coveredCol)];
                        ++renderStats_.changedCells;
                    }
                }
                col += clearWidth;
                renderStats_.changedRuns += static_cast<std::size_t>(clearHeight + 1);
                continue;
            }
            std::string run;
            int runCol = col;
            while (col < lastCol) {
                const std::size_t currentIndex = index(row, col);
                const TerminalCell& current = cells_[currentIndex];
                if (!fullRedraw && current == previous_[currentIndex]) {
                    break;
                }
                if (current.rawSkip) {
                    break;
                }
                if (!current.raw.empty()) {
                    break;
                }
                if (current.foreground != styleCell.foreground ||
                    current.background != styleCell.background) {
                    break;
                }
                run += terminalCellText(current.text);
                previous_[currentIndex] = current;
                ++col;
                ++renderStats_.changedCells;
            }

            output += "\x1b[" + std::to_string(row + rowOffset + kAnsiBaseRow) + ";" +
                      std::to_string(runCol + colOffset + kAnsiBaseCol) + "H";
            output += sgrForCell(styleCell);
            output += run;
            ++renderStats_.changedRuns;
        }
    }
    if (!output.empty()) {
        if (rawEmitted) {
            for (int row = firstRow; row < lastRow; ++row) {
                int col = firstCol;
                while (col < lastCol) {
                    const TerminalCell& cell = cells_[index(row, col)];
                    if (cell.rawSkip || !cell.raw.empty()) {
                        ++col;
                        continue;
                    }
                    const TerminalCell styleCell = cell;
                    const int runCol = col;
                    std::string run;
                    while (col < lastCol) {
                        const TerminalCell& current = cells_[index(row, col)];
                        if (current.rawSkip || !current.raw.empty()) {
                            break;
                        }
                        if (current.foreground != styleCell.foreground ||
                            current.background != styleCell.background) {
                            break;
                        }
                        run += terminalCellText(current.text);
                        ++col;
                    }
                    output += "\x1b[" + std::to_string(row + rowOffset + kAnsiBaseRow) + ";" +
                              std::to_string(runCol + colOffset + kAnsiBaseCol) + "H";
                    output += sgrForCell(styleCell);
                    output += run;
                    ++renderStats_.changedRuns;
                }
            }
        }
        output += "\x1b[0m";
    }
    if (!output.empty() && synchronizeUpdate) {
        output = std::string{kAnsiSyncUpdateBegin} + output + kAnsiSyncUpdateEnd;
    }
    renderStats_.outputBytes += output.size();
    forceFullRedraw_ = false;
    return output;
}

std::string TerminalBuffer::renderScrollRegion(int rowOffset, int startRow, int height, int delta) {
    const int firstRow = std::max(0, startRow);
    const int lastRow = std::min(height_, startRow + std::max(0, height));
    const int regionHeight = lastRow - firstRow;
    const int distance = std::min(std::abs(delta), regionHeight);
    if (forceFullRedraw_ || regionHeight <= 1 || distance <= 0 || distance >= regionHeight) {
        return "";
    }

    const std::vector<TerminalCell> before = previous_;
    if (delta > 0) {
        for (int row = lastRow - 1; row >= firstRow + distance; --row) {
            for (int col = 0; col < width_; ++col) {
                previous_[index(row, col)] = before[index(row - distance, col)];
            }
        }
        for (int row = firstRow; row < firstRow + distance; ++row) {
            for (int col = 0; col < width_; ++col) {
                previous_[index(row, col)] = TerminalCell{};
            }
        }
    } else {
        for (int row = firstRow; row < lastRow - distance; ++row) {
            for (int col = 0; col < width_; ++col) {
                previous_[index(row, col)] = before[index(row + distance, col)];
            }
        }
        for (int row = lastRow - distance; row < lastRow; ++row) {
            for (int col = 0; col < width_; ++col) {
                previous_[index(row, col)] = TerminalCell{};
            }
        }
    }

    const char command = delta > 0 ? 'T' : 'S';
    std::string output =
        "\x1b[" + std::to_string(firstRow + rowOffset + kAnsiBaseRow) + ";" +
        std::to_string(lastRow + rowOffset) + "r" +
        "\x1b[" + std::to_string(firstRow + rowOffset + kAnsiBaseRow) + ";1H" +
        "\x1b[" + std::to_string(distance) + command +
        "\x1b[r";
    renderStats_.outputBytes += output.size();
    return output;
}

std::size_t TerminalBuffer::index(int row, int col) const {
    return static_cast<std::size_t>(row * width_ + col);
}

bool TerminalBuffer::inBounds(int row, int col) const {
    return row >= 0 && row < height_ && col >= 0 && col < width_;
}

}  // namespace ui

File: java/src/main/java/uimd/TerminalBuffer.java
package uimd;

import java.nio.charset.StandardCharsets;

public final class TerminalBuffer
{
    private static final int ANSI_BASE_ROW = 1;
    private static final int ANSI_BASE_COL = 1;
    private static final int MINIMUM_SCROLL_REGION_ROWS = 2;
    private static final String ANSI_RESET = "\u001b[0m";
    private static final String ANSI_SYNC_UPDATE_BEGIN = "\u001b[?2026h";
    private static final String ANSI_SYNC_UPDATE_END = "\u001b[?2026l";
    private static final String ANSI_RESET_SCROLL_REGION = "\u001b[r";

    private int width;
    private int height;
    private TerminalCell[][] cells;
    private TerminalCell[][] previous;
    private boolean forceFullRedraw = true;
    private TerminalBufferRenderStats renderStats = new TerminalBufferRenderStats(0, 0, 0, false);

    public TerminalBuffer(int width, int height)
    {
        resizeStorage(width, height);
    }

    public int width()
    {
        return width;
    }

    public int height()
    {
        return height;
    }

    public void resize(int newWidth, int newHeight)
    {
        validateDimensions(newWidth, newHeight);
        if (newWidth == width && newHeight == height)
        {
            return;
        }
        resizeStorage(newWidth, newHeight);
    }

    public void clear()
    {
        clear(new TerminalCell());
    }

    public void clear(TerminalCell cell)
    {
        TerminalCell fill = cell == null ? new TerminalCell() : cell;
        for (int row = 0; row < height; ++row)
        {
            for (int col = 0; col < width; ++col)
            {
                cells[row][col] = fill.copy();
            }
        }
    }

    public TerminalCell cell(int row, int col)
    {
        if (!inBounds(row, col))
        {
            throw new IndexOutOfBoundsException("terminal cell is outside the buffer");
        }
        return cells[row][col];
    }

    public void setCell(int row, int col, TerminalCell cell)
    {
        if (inBounds(row, col))
        {
            cells[row][col] = cell == null ? new TerminalCell() : cell;
        }
    }

    public void requestFullRedraw()
    {
        forceFullRedraw = true;
    }

    public void beginRenderStats()
    {
        renderStats = new TerminalBufferRenderStats(0, 0, 0, false);
    }

    public TerminalBufferRenderStats renderStats()
    {
        return renderStats;
    }

    public String renderDiff()
    {
        return renderDiff(0, 0);
    }

    public String renderDiff(int rowOffset, int colOffset)
    {
        return renderDiffRegion(rowOffset, colOffset, 0, 0, height, width);
    }

    public String renderDiffRegion(
        int rowOffset,
        int colOffset,
        int startRow,
        int startCol,
        int regionHeight,
        int regionWidth)
    {
        StringBuilder output = new StringBuilder();
        boolean fullRedraw = forceFullRedraw;
        boolean synchronizedUpdate = false;
        boolean rawEmitted = false;
        long changedCells = 0;
        long changedRuns = 0;
        int firstRow = Math.max(0, startRow);
        int firstCol = Math.max(0, startCol);
        int lastRow = Math.min(height, startRow + Math.max(0, regionHeight));
        int lastCol = Math.min(width, startCol + Math.max(0, regionWidth));

        for (int row = firstRow; row < lastRow; ++row)
        {
            int col = firstCol;
            while (col < lastCol)
            {
                TerminalCell start = cells[row][col];
                if (start.rawSkip())
                {
                    if (fullRedraw || !start.equals(previous[row][col]))
                    {
                        ++changedCells;
                    }
                    previous[row][col] = start.copy();
                    ++col;
                    continue;
                }
                if (!fullRedraw && start.equals(previous[row][col]))
                {
                    ++col;
                    continue;
                }

                TerminalCell styleCell = start;
                if (!styleCell.raw().isEmpty())
                {
                    synchronizedUpdate = true;
                    int rawWidth = Math.max(1, styleCell.rawWidth());
                    int rawHeight = Math.max(1, styleCell.rawHeight());
                    int clearWidth = Math.min(rawWidth, width - col);
                    int clearHeight = Math.min(rawHeight, height - row);
                    for (int clearRow = row; clearRow < row + clearHeight; ++clearRow)
                    {
                        output.append(cursorPosition(clearRow + rowOffset, col + colOffset));
                        output.append(RenderHelpers.sgrForCell(styleCell));
                        output.append(" ".repeat(clearWidth));
                    }
                    if (clearHeight >= rawHeight)
                    {
                        int anchorRow = row + rowOffset;
                        String noScrollRegion = rawNoScrollRegion(anchorRow, rawHeight, rowOffset + height);
                        output.append(noScrollRegion);
                        output.append(cursorPosition(row + rowOffset, col + colOffset));
                        output.append(styleCell.raw());
                        if (!noScrollRegion.isEmpty())
                        {
                            output.append(ANSI_RESET_SCROLL_REGION);
                        }
                        rawEmitted = true;
                    }
                    for (int coveredRow = row; coveredRow < row + clearHeight; ++coveredRow)
                    {
                        for (int coveredCol = col; coveredCol < col + clearWidth; ++coveredCol)
                        {
                            previous[coveredRow][coveredCol] = cells[coveredRow][coveredCol].copy();
                            ++changedCells;
                        }
                    }
                    changedRuns += clearHeight + 1L;
                    col += clearWidth;
                    continue;
                }

                StringBuilder run = new StringBuilder();
                int runCol = col;
                while (col < lastCol)
                {
                    TerminalCell current = cells[row][col];
                    if (!fullRedraw && current.equals(previous[row][col]))
                    {
                        break;
                    }
                    if (current.rawSkip() || !current.raw().isEmpty())
                    {
                        break;
                    }
                    if (!RenderHelpers.sameColor(current.foreground(), styleCell.foreground())
                        || !RenderHelpers.sameColor(current.background(), styleCell.background()))
                    {
                        break;
                    }
                    run.append(RenderHelpers.safeTerminalCellText(current.text()));
                    previous[row][col] = current.copy();
                    ++changedCells;
                    ++col;
                }
                output.append(cursorPosition(row + rowOffset, runCol + colOffset));
                output.append(RenderHelpers.sgrForCell(styleCell));
                output.append(run);
                ++changedRuns;
            }
        }

        if (!output.isEmpty())
        {
            if (rawEmitted)
            {
                changedRuns += repaintTextOverRaw(
                    output,
                    rowOffset,
                    colOffset,
                    firstRow,
                    firstCol,
                    lastRow,
                    lastCol);
            }
            output.append(ANSI_RESET);
        }
        forceFullRedraw = false;
        String result = output.toString();
        if (!result.isEmpty() && synchronizedUpdate)
        {
            result = ANSI_SYNC_UPDATE_BEGIN + result + ANSI_SYNC_UPDATE_END;
        }
        renderStats = new TerminalBufferRenderStats(
            renderStats.changedCells() + changedCells,
            renderStats.changedRuns() + changedRuns,
            renderStats.outputBytes() + result.getBytes(StandardCharsets.UTF_8).length,
            renderStats.fullRedraw() || fullRedraw);
        return result;
    }

    public String renderScrollRegion(int rowOffset, int startRow, int regionHeight, int delta)
    {
        int firstRow = Math.max(0, startRow);
        int lastRow = Math.min(height, startRow + Math.max(0, regionHeight));
        int actualHeight = lastRow - firstRow;
        int distance = Math.min(Math.abs(delta), actualHeight);
        if (forceFullRedraw || actualHeight <= 1 || distance <= 0 || distance >= actualHeight)
        {
            return "";
        }
        if (scrollRegionHasRawCells(firstRow, lastRow))
        {
            return "";
        }

        TerminalCell[][] before = copyGrid(previous);
        if (delta > 0)
        {
            for (int row = lastRow - 1; row >= firstRow + distance; --row)
            {
                for (int col = 0; col < width; ++col)
                {
                    previous[row][col] = before[row - distance][col].copy();
                }
            }
            blankPreviousRows(firstRow, firstRow + distance);
        }
        else
        {
            for (int row = firstRow; row < lastRow - distance; ++row)
            {
                for (int col = 0; col < width; ++col)
                {
                    previous[row][col] = before[row + distance][col].copy();
                }
            }
            blankPreviousRows(lastRow - distance, lastRow);
        }
        char command = delta > 0 ? 'T' : 'S';
        String output = "\u001b[" + (firstRow + rowOffset + ANSI_BASE_ROW) + ";" + (lastRow + rowOffset) + "r"
            + "\u001b[" + (firstRow + rowOffset + ANSI_BASE_ROW) + ";1H"
            + "\u001b[" + distance + command
            + ANSI_RESET_SCROLL_REGION;
        renderStats = new TerminalBufferRenderStats(
            renderStats.changedCells(),
            renderStats.changedRuns(),
            renderStats.outputBytes() + output.getBytes(StandardCharsets.UTF_8).length,
            renderStats.fullRedraw());
        return output;
    }

    private void resizeStorage(int newWidth, int newHeight)
    {
        validateDimensions(newWidth, newHeight);
        width = newWidth;
        height = newHeight;
        cells = new TerminalCell[height][width];
        previous = new TerminalCell[height][width];
        for (int row = 0; row < height; ++row)
        {
            for (int col = 0; col < width; ++col)
            {
                cells[row][col] = new TerminalCell();
                previous[row][col] = new TerminalCell();
            }
        }
        forceFullRedraw = true;
        beginRenderStats();
    }

    private boolean inBounds(int row, int col)
    {
        return row >= 0 && row < height && col >= 0 && col < width;
    }

    private void blankPreviousRows(int firstRow, int lastRow)
    {
        for (int row = firstRow; row < lastRow; ++row)
        {
            for (int col = 0; col < width; ++col)
            {
                previous[row][col] = new TerminalCell();
            }
        }
    }

    private boolean scrollRegionHasRawCells(int firstRow, int lastRow)
    {
        for (int row = firstRow; row < lastRow; ++row)
        {
            for (int col = 0; col < width; ++col)
            {
                TerminalCell current = cells[row][col];
                TerminalCell prior = previous[row][col];
                if (!current.raw().isEmpty() || current.rawSkip()
                    || !prior.raw().isEmpty() || prior.rawSkip())
                {
                    return true;
                }
            }
        }
        return false;
    }

    private long repaintTextOverRaw(
        StringBuilder output,
        int rowOffset,
        int colOffset,
        int firstRow,
        int firstCol,
        int lastRow,
        int lastCol)
    {
        long changedRuns = 0;
        for (int row = firstRow; row < lastRow; ++row)
        {
            int col = firstCol;
            while (col < lastCol)
            {
                TerminalCell cell = cells[row][col];
                if (cell.rawSkip() || !cell.raw().isEmpty())
                {
                    ++col;
                    continue;
                }
                TerminalCell styleCell = cell;
                int runCol = col;
                StringBuilder run = new StringBuilder();
                while (col < lastCol)
                {
                    TerminalCell current = cells[row][col];
                    if (current.rawSkip() || !current.raw().isEmpty()
                        || !RenderHelpers.sameColor(current.foreground(), styleCell.foreground())
                        || !RenderHelpers.sameColor(current.background(), styleCell.background()))
                    {
                        break;
                    }
                    run.append(RenderHelpers.safeTerminalCellText(current.text()));
                    ++col;
                }
                output.append(cursorPosition(row + rowOffset, runCol + colOffset));
                output.append(RenderHelpers.sgrForCell(styleCell));
                output.append(run);
                ++changedRuns;
            }
        }
        return changedRuns;
    }

    private static String cursorPosition(int row, int col)
    {
        return "\u001b[" + (row + ANSI_BASE_ROW) + ";" + (col + ANSI_BASE_COL) + "H";
    }

    private static String rawNoScrollRegion(int anchorRow, int rawHeight, int bufferBottomExclusive)
    {
        if (anchorRow >= MINIMUM_SCROLL_REGION_ROWS)
        {
            return "\u001b[1;" + anchorRow + "r";
        }
        int rawBottomExclusive = anchorRow + Math.max(1, rawHeight);
        if (bufferBottomExclusive - rawBottomExclusive >= MINIMUM_SCROLL_REGION_ROWS)
        {
            return "\u001b[" + (rawBottomExclusive + ANSI_BASE_ROW) + ";" + bufferBottomExclusive + "r";
        }
        return "";
    }

    private static TerminalCell[][] copyGrid(TerminalCell[][] source)
    {
        TerminalCell[][] result = new TerminalCell[source.length][];
        for (int row = 0; row < source.length; ++row)
        {
            result[row] = new TerminalCell[source[row].length];
            for (int col = 0; col < source[row].length; ++col)
            {
                result[row][col] = source[row][col].copy();
            }
        }
        return result;
    }

    private static void validateDimensions(int width, int height)
    {
        if (width < 0 || height < 0)
        {
            throw new IllegalArgumentException("terminal buffer dimensions must not be negative");
        }
    }
}
